core/src/execution/datafusion/expressions/utils.rs

```diff
@@ -134,36 +140,40 @@ pub fn array_with_timezone(
 ///     array - input array of timestamp without timezone
 ///     tz - timezone of the values in the input array
 ///     to_timezone - timezone to change the input values to
-fn timestamp_ntz_to_timestamp(array: ArrayRef, tz: &str, to_timezone: Option<&str>) -> ArrayRef {
+fn timestamp_ntz_to_timestamp(
+    array: ArrayRef,
+    tz: &str,
+    to_timezone: Option<&str>,
+) -> Result<ArrayRef, ArrowError> {
     assert!(!tz.is_empty());
     match array.data_type() {
         DataType::Timestamp(_, None) => {
             let array = as_primitive_array::<TimestampMicrosecondType>(&array);
-            let tz: Tz = tz.parse().unwrap();
-            let values = array.iter().map(|v| {
-                v.map(|value| {
-                    let local_datetime = as_datetime::<TimestampMicrosecondType>(value).unwrap();
-                    let datetime: DateTime<Tz> = tz.from_local_datetime(&local_datetime).unwrap();
-                    datetime.timestamp_micros()
-                })
-            });
-            let mut array: PrimitiveArray<TimestampMicrosecondType> =
-                unsafe { PrimitiveArray::from_trusted_len_iter(values) };
-            array = if let Some(to_tz) = to_timezone {
+            let tz: Tz = tz.parse()?;
```

Comment:
We should probably add a case for an invalid timezone in the unit test.

Reply:
Added in cdbf87ae47a4e2b12d76f08138105340097d00f3